### PR TITLE
handling hosts and files sanely

### PR DESCRIPTION
**Pull request by larsks**
**DO NOT MERGE**

this change introduces database schema changes necessary to support
both #102 and #103:

- Hosts are now only unique within a playbook run (so you can have
  multiple hosts with the same name in the hosts table as long as they
  are associated with different top-level playbooks).

- Files are treated the same way.  Each task is linked to a file, and
  filenames are unique within a playbook run, but we don't assume that
  /path/to/your/tasks.yml in one run is the same in another run.

This includes the necessary changes to the callback plugin to support
the new schema.

**This commit does not contain web ui updates**

### CODE REVIEW

**Comment by larsks** [comment]
...or test updates, at this point.

File: ara/callback/log_ara.py [hunk 1]
```diff
@@ -85,13 +85,28 @@ def __init__(self):
...
         except models.NoResultFound:
-            host = models.Host(name=hostname)
+            host = models.Host(name=hostname, playbook=self.playbook)
             db.session.add(host)
 
         return host
 
+    def get_or_create_file(self, path):
+        try:
+            file_ = (models.File.query
```
Comment (dmsimard) [inline ara/callback/log_ara.py]: Any particular reason for the underscore ?

File: ara/cli/file.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#   License for the specific language governing permissions and limitations
+#   under the License.
+#
+
+import json
+import logging
+import six
```
Comment (dmsimard) [inline ara/cli/file.py]: It doesn't look like six is used here.

File: ara/cli/file.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+class FileShow(ShowOne):
+    """Show details of a file"""
+    log = logging.getLogger(__name__)
+
+    def get_parser(self, prog_name):
+        parser = super(FileShow, self).get_parser(prog_name)
+        parser.add_argument(
+            'file',
+            metavar='<host>',
```
Comment (dmsimard) [inline ara/cli/file.py]: wrong metavar

File: ara/callback/log_ara.py (hunk 1 above)
Comment (larsks) [inline ara/callback/log_ara.py]: Because `file` is a built-in Python type (that's what, for example, the `open` command returns), and I try avoid masking those.

File: ara/cli/file.py (hunk 2 above)
Comment (larsks) [inline ara/cli/file.py]: Yeah, probably not, but that will turn up in flake8 testing.  I am more interesting in the architecture of this pr than in things like this that will be flushed out by our automated tools...

File: ara/callback/log_ara.py [hunk 4]
```diff
@@ -85,13 +87,39 @@ def __init__(self):
...
+                     .filter_by(path=path)
+                     .filter_by(playbook_id=self.playbook.id)
+                     .one())
+        except models.NoResultFound:
+            file_ = models.File(path=path, playbook=self.playbook)
+            db.session.add(file_)
+
+            # LKS: should collecting content be optional?
```
Comment (dmsimard) [inline ara/callback/log_ara.py]: We could parameterize it but default to true IMO.

File: ara/callback/log_ara.py [hunk 5]
```diff
@@ -85,13 +87,39 @@ def __init__(self):
...
+            with open(path, 'r') as fd:
+                sha1 = hashlib.sha1(fd.read()).hexdigest()
```
Comment (dmsimard) [inline ara/callback/log_ara.py]: Perhaps the method to calculate the sha1 should be pulled from models (no code dedup but especially less chance for regression)

File: ara/models.py [hunk 6]
```diff
@@ -45,11 +55,17 @@ def mkuuid():
     db.relationship, passive_deletes=False,
     cascade='all, delete-orphan', uselist=False)
 
-# Common options for one-to-many relationships in our database.
+# common options for one-to-many relationships in our database.
 one_to_many = functools.partial(
     db.relationship, passive_deletes=False,
     cascade='all, delete-orphan', lazy='dynamic')
 
+# common options for one-to-many relationships in our database.
```
Comment (dmsimard) [inline ara/models.py]: typo one-to-many

File: ara/models.py [hunk 7]
```diff
@@ -114,6 +133,66 @@ def __repr__(self):
...
+    # restrictions stems from the fact that we are using this column in
+    # a UNIQUE constraint.
+    path = db.Column(db.String(3000))
+    content = many_to_one('FileContent', backref='files')
+    content_id = db.Column(db.String(40),
+                           db.ForeignKey('file_contents.id'))
+
+    # is_primary is true for playbooks referenced directly on the
+    # ansible-playbook command line.
+    is_primary = db.Column(db.Boolean, default=False)
```
Comment (dmsimard) [inline ara/models.py]: warning: bikeshedding inc: is_playbook instead ?

File: ara/models.py [hunk 8]
```diff
@@ -114,6 +133,66 @@ def __repr__(self):
...
+
+
+class CompressedText(types.TypeDecorator):
```
Comment (dmsimard) [inline ara/models.py]: This is awesome in a lot of ways.

Comment (dmsimard) [inline ara/models.py]: We should use this for the result field which can also be large.

File: ara/cli/file.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        return [[field.name for field in FIELDS],
+                [[field(file_) for field in FIELDS]
+                 for file_ in files]]
+
+
+class FileShow(ShowOne):
+    """Show details of a file"""
+    log = logging.getLogger(__name__)
+
+    def get_parser(self, prog_name):
```
Comment (dmsimard) [inline ara/cli/file.py]: Could probably add a --long format to also show the file contents but that can be added later

File: ara/callback/log_ara.py (hunk 5 above)
Comment (larsks) [inline ara/callback/log_ara.py]: Can you elaborate on that?  I'm not sure I follow.  I mean, the model needs to know how to calculate the content hash...

File: ara/models.py (hunk 7 above)
Comment (larsks) [inline ara/models.py]: That sounds fine to me.

File: ara/cli/host.py [hunk 10]
```diff
@@ -62,19 +78,17 @@ def get_parser(self, prog_name):
...
         )
         return parser
 
     def take_action(self, args):
-        try:
-            host = (models.Host.query
-                    .filter((models.Host.id == args.host) |
-                            (models.Host.name == args.host)).one())
-        except models.NoResultFound:
+        host = models.Host.query.get(args.host)
```
Comment (dmsimard) [inline ara/cli/host.py]: The way OSC handles this is that if the search by name returns len > 1, they raise an exception. i.e: https://github.com/openstack/python-openstackclient/blob/master/openstackclient/common/utils.py#L100

We could keep that feature and handle it similarly for fields that are possibly non-unique

File: ara/cli/playbook.py [hunk 11]
```diff
@@ -133,13 +145,5 @@ def take_action(self, args):
             self.log.warning('deleting playbook %s', pid)
             playbook = models.Playbook.query.get(pid)
             db.session.delete(playbook)
-            orphan_hosts = (models.Host.query
```
Comment (dmsimard) [inline ara/cli/playbook.py]: So this part is now handled right in the model with the relationships ?

File: ara/cli/host.py (hunk 10 above)
Comment (larsks) [inline ara/cli/host.py]: I'm not sure I see the advantage there.  I mean, calling `.one()` does exactly that (raises an exception if more than one result is found).

File: ara/templates/navbar.html [hunk 12]
```diff
@@ -12,16 +12,6 @@
     <div class="navbar-collapse collapse">
       <ul class="nav navbar-nav">
         <li class="dropdown">
-          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Hosts <span class="caret"></span></a>
```
Comment (dmsimard) [inline ara/templates/navbar.html]: So we'll only show playbooks then ? This is fine but I think this means we could probably do away with the navbar entirely and do a nicely featured home page like I mentioned here: https://github.com/dmsimard/ara/issues/76#issuecomment-222329270

Can be done in a follow up and I can take a stab at it.

File: ara/models.py (hunk 8 above)
Comment (larsks) [inline ara/models.py]: That...is a fantastic idea.

File: ara/callback/log_ara.py (hunk 5 above)
Comment (dmsimard) [inline ara/callback/log_ara.py]: Instead of re-doing "hashlib.sha1()" here, we could use "models.content_sha1()"

File: ara/cli/host.py (hunk 10 above)
Comment (dmsimard) [inline ara/cli/host.py]: It removes the advantage of being able to search by name

Comment (larsks) [inline ara/cli/host.py]: Oh, I see. That's different.  I was thinking that we could also support `host show <playbook> <hostname>`.  I think that something like `host show <hostname>` would always fail (because if you run a playbook more than once, you will have multiple matches for `<hostname>`).

Comment (larsks) [inline ara/cli/host.py]: I guess maybe we could have something like `host show --latest <hostname>`...

File: ara/templates/navbar.html (hunk 12 above)
Comment (larsks) [inline ara/templates/navbar.html]: Yeah, I was thinking that in general we need to substantially re-architect the UI.  The filter-by-file stuff I put in is really just a stopgap measure; I'm sure there's a much nicer way to handle that, and of course we also want to include tools for exposing the file content now that we are collecting that.

File: ara/cli/result.py [hunk 13]
```diff
@@ -89,8 +114,17 @@ def get_parser(self, prog_name):
         parser = super(ResultShow, self).get_parser(prog_name)
         parser.add_argument(
```
Comment (larsks) [inline ara/cli/result.py]: Makes sense.

File: ara/cli/playbook.py (hunk 11 above)
Comment (larsks) [inline ara/cli/playbook.py]: Right.

File: ara/cli/host.py (hunk 10 above)
Comment (dmsimard) [inline ara/cli/host.py]: Oh, that's right, one host can represent multiple host rows because it is unique per playbook id, not unique in the hosts table. Hmmm.. Let's leave it like this for the time being and see if we come up with a better UX later.

File: ara/templates/playbook.html [hunk 14]
```diff
@@ -56,6 +56,7 @@
 <table class="plays table table-striped table-bordered table-hover table-condensed">
   <thead>
     <tr>
+      <t>Name</th>
```
Comment (dmsimard) [inline ara/templates/playbook.html]: <th> typo

File: ara/templates/playbook.html [hunk 15]
```diff
@@ -93,8 +106,9 @@
     <tr>
       <td>{{ task.offset_from_playbook |timefmt }}</td>
       <td>{{ task.duration |timefmt }}</td>
+      <td>{{ task.name }}</td>
       <td>{{ task.action }}</td>
-      <td>{{ task.path|pathtruncate }}</td>
+      <td>{{ task.file.path|pathtruncate }}</td>
```
Comment (dmsimard) [inline ara/templates/playbook.html]: Could probably make this clickable and lead to the file content/task list/etc ?

File: ara/templates/playbook.html [hunk 16]
```diff
@@ -66,21 +67,33 @@
...
 <h2>Tasks</h2>
+{% if not file_ %}
+<p>This is a list of all tasks encounted in the Ansible run.  You may
+limit this to a list of tasks contained in a particular file by going
+to {{ macros.make_link('playbook.show_playbook_files', 'the list of files',
+playbook=playbook.id) }}.</p>
+{% else %}
+<p>This is a list of tasks in <code>{{ file_.path }}</code>.  You may
+also see {{ macros.make_link('playbook.show_playbook',
+'the complete list of tasks', playbook=playbook.id) }}.</p>
```
Comment (dmsimard) [inline ara/templates/playbook.html]: Nit but there's a space before the period in the UI

Alt

File: ara/templates/playbook.html [hunk 17]
```diff
@@ -93,8 +106,9 @@
     <tr>
       <td>{{ task.offset_from_playbook |timefmt }}</td>
       <td>{{ task.duration |timefmt }}</td>
+      <td>{{ task.name }}</td>
       <td>{{ task.action }}</td>
-      <td>{{ task.path|pathtruncate }}</td>
+      <td>{{ task.file.path|pathtruncate }}</td>
       <td>{{ task.lineno }}</td>
```
Comment (dmsimard) [inline ara/templates/playbook.html]: I have no clue how to do that but I could definitely see this line link to the file contents page with the line highlighted (similar to github)

File: ara/templates/playbook_file_summary.html [hunk 18]
```diff
@@ -0,0 +1,26 @@
+{% extends "layout.html" %}
```
Comment (dmsimard) [inline ara/templates/playbook_file_summary.html]: the fact that a file is primary (or playbook) should probably be highlighted somehow

Comment (larsks) [inline ara/templates/playbook_file_summary.html]: I can add another column, unless you have a better idea...

File: ara/templates/playbook.html (hunk 17 above)
Comment (larsks) [inline ara/templates/playbook.html]: That is what I want to have, ultimately, but I was going to leave that for another pr, since I think that's going to be e more substantial chunk of UI work.

File: ara/callback/log_ara.py (hunk 5 above)
Comment (larsks) [inline ara/callback/log_ara.py]: We can't though; that function only exists because we're using it as the default for a sqlalchemy column attribute, and the code is highly sqlalchemy specific.  We can't easily just pass it arbitrary content.  I could *further* abstract the function by wrapping the call to hashalib.sha1 in something else, but I'm not sure that gets us anything.

File: ara/templates/playbook_file_summary.html (hunk 18 above)
Comment (dmsimard) [inline ara/templates/playbook_file_summary.html]: I think there's only ever going to be one primary file ? For a given playbook, that file can be at the top (outside the table or something) and then the task/role files below. It'd be kind of silly to add another column that'll really only be "ticked" for the primary file.

File: ara/templates/playbook.html (hunk 17 above)
Comment (dmsimard) [inline ara/templates/playbook.html]: Yes, another PR. Just thinking out loud. I'll formalize it in an issue.

File: ara/callback/log_ara.py (hunk 5 above)
Comment (dmsimard) [inline ara/callback/log_ara.py]: I don't have a strong opinion. If we feel we're going to be using the sha1 call in too many places, it should go in utils. This is just two places for the time being (model and here) but will it be used elsewhere as well ?

**Comment by larsks** [comment]
I've pushed a new set of changes that I think includes all the changes you suggested, except that I'm still using `hashlib.sha1` in two places.  Moving it into `utils` might make sense, but including `utils` from `models` creates a circular dependency that I don't want to deal with right now.

File: ara/cli/host.py [hunk 19]
```diff
@@ -76,15 +76,28 @@ class HostShow(ShowOne):
...
         parser.add_argument(
+            '--playbook', '-b',
+            metavar='<playbook-id>',
+            help='Show host associated with the given playbook',
+        )
+        parser.add_argument(
             'host',
             metavar='<host>',
-            help='Host id to show',
+            help='Host name (when using --playbook) or id to show',
```
Comment (dmsimard) [inline ara/cli/host.py]: I think this could be better phrased "Host id (or name when using --playbook) to show".

**Comment by dmsimard** [comment]
Closing, we have moved to review.openstack.org.
